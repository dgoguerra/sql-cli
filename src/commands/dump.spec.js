const fs = require("fs");
const path = require("path");
const tar = require("tar");
const rimraf = require("rimraf");
const { getTestKnex, runCli } = require("../../test/utils");
const { hydrateKnex } = require("../knexUtils");

const TEST_DUMP_NAME = "my_dump";
const TEST_DUMP_PATH = `${process.env.PWD}/.tmp/${TEST_DUMP_NAME}.tgz`;
const TEST_EXTRACTED_PATH = `${process.env.PWD}/.tmp/${TEST_DUMP_NAME}`;
const TEST_MIGRATIONS_TABLE = `migrations_${TEST_DUMP_NAME}`;

const TEST_DATETIME_1 = "2020-07-24 18:34:00";
const TEST_DATETIME_2 = "2020-07-24 19:25:00";

const TEST_TABLE1_CONTENT = [
  { field_1: 12, field_2: "foo", created_at: TEST_DATETIME_1 },
  { field_1: 30, field_2: "bar", created_at: TEST_DATETIME_1 },
];

const TEST_TABLE2_CONTENT = [
  { field_1: 12.3, field_2: "foo", created_at: TEST_DATETIME_1 },
  { field_1: 30.45, field_2: "bar", created_at: TEST_DATETIME_2 },
];

describe("dump", () => {
  let knex;
  let connUri;

  let knex2;
  let connUri2;

  beforeAll(async () => {
    knex = hydrateKnex(getTestKnex());
    connUri = knex.getUri();

    knex2 = hydrateKnex(getTestKnex("test-load.db"));
    connUri2 = knex2.getUri();

    await knex.schema.createTable("table_1", (t) => {
      t.increments("id");
      t.integer("field_1");
      t.text("field_2");
      t.timestamps();
      t.index(["field_1"]);
    });
    await knex.schema.createTable("table_2", (t) => {
      t.increments("id");
      t.decimal("field_1").notNullable();
      t.text("field_2").defaultTo("default text");
      t.timestamps();
      t.unique(["field_1"]);
    });
    await knex.schema.createTable("table_3", (t) => {
      t.bigIncrements("id_field");
      t.bigInteger("field_1");
      t.timestamps();
    });

    await knex("table_1").insert(TEST_TABLE1_CONTENT);
    await knex("table_2").insert(TEST_TABLE2_CONTENT);

    rimraf.sync(TEST_DUMP_PATH);
    rimraf.sync(TEST_EXTRACTED_PATH);
  });

  it("can create database dump", async () => {
    expect(fs.existsSync(TEST_DUMP_PATH)).toBeFalsy();
    await runCli(`dump create ${connUri} ${TEST_DUMP_PATH}`);
    expect(fs.existsSync(TEST_DUMP_PATH)).toBeTruthy();
  });

  it("dumpfile contents are valid", async () => {
    expect(fs.existsSync(TEST_EXTRACTED_PATH)).toBeFalsy();

    await tar.extract({
      file: TEST_DUMP_PATH,
      cwd: path.dirname(TEST_DUMP_PATH),
    });
    expect(fs.existsSync(TEST_EXTRACTED_PATH)).toBeTruthy();

    const files = [
      "migrations/20200722182250-table_1.js",
      "migrations/20200722182250-table_2.js",
      "migrations/20200722182250-table_3.js",
      "data/table_1.jsonl",
      "data/table_2.jsonl",
    ];
    files.forEach((file) => {
      const filePath = `${TEST_EXTRACTED_PATH}/${file}`;
      expect(fs.existsSync(filePath)).toBeTruthy();
      expect(fs.readFileSync(filePath).toString()).toMatchSnapshot();
    });
  });

  it("can see dumpfile contents", async () => {
    expect(await runCli(`ls ${TEST_DUMP_PATH}`)).toMatchSnapshot();

    expect(await runCli(`show ${TEST_DUMP_PATH}/table_1`)).toMatchSnapshot();

    expect(await runCli(`diff ${connUri} ${TEST_DUMP_PATH}`)).toMatchSnapshot();

    expect(
      await runCli(`diff ${connUri}/table_1 ${TEST_DUMP_PATH}/table_1`)
    ).toMatchSnapshot();
  });

  it("can load dump to database", async () => {
    expect(await knex2.schema.listTables()).toMatchObject([]);

    await runCli(`dump load ${connUri2} ${TEST_DUMP_PATH}`);

    expect(await knex2.schema.listTables()).toMatchObject([
      { table: TEST_MIGRATIONS_TABLE },
      { table: `${TEST_MIGRATIONS_TABLE}_lock` },
      { table: "table_1" },
      { table: "table_2" },
      { table: "table_3" },
    ]);
  });

  it("loaded database has expected content", async () => {
    expect(await knex2(TEST_MIGRATIONS_TABLE)).toMatchObject([
      { id: 1, batch: 1, name: "20200722182250-table_1.js" },
      { id: 2, batch: 1, name: "20200722182250-table_2.js" },
      { id: 3, batch: 1, name: "20200722182250-table_3.js" },
    ]);
    expect(await knex2(`${TEST_MIGRATIONS_TABLE}_lock`)).toMatchObject([
      { index: 1, is_locked: 0 },
    ]);

    expect(await runCli(`show ${connUri2}/table_1`)).toMatchSnapshot();
    expect(await runCli(`show ${connUri2}/table_2`)).toMatchSnapshot();
    expect(await runCli(`show ${connUri2}/table_3`)).toMatchSnapshot();

    // sqlite returns dates as their numeric value in milliseconds
    const toSqliteRow = (row) => {
      row.created_at = new Date(row.created_at).getTime();
      return row;
    };

    expect(await knex2("table_1")).toMatchObject(
      TEST_TABLE1_CONTENT.map(toSqliteRow)
    );
    expect(await knex2("table_2")).toMatchObject(
      TEST_TABLE2_CONTENT.map(toSqliteRow)
    );
    expect(await knex2("table_3")).toMatchObject([]);
  });

  it("can diff against original database", async () => {
    expect(await runCli(`diff ${connUri} ${connUri2}`)).toMatchSnapshot();

    expect(
      await runCli(`diff ${connUri}/table_1 ${connUri2}/table_1`)
    ).toMatchSnapshot();
  });
});
