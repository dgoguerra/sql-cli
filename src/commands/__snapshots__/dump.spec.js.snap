// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dump can diff against original database 1`] = `
"table                   rows bytes  columns    indexes
migrations_my_dump      3    4.1 kB 4x created none
migrations_my_dump_lock 1    4.1 kB 2x created none

Tables: 2x created, 3x similar (hidden)
Re-run with --all to show hidden rows
"
`;

exports[`dump can diff against original database 2`] = `
"Columns: 5x similar (hidden)
Indexes: 1x similar (hidden)
Re-run with --all to show hidden rows
"
`;

exports[`dump can see dumpfile contents 1`] = `
"table                   rows bytes
migrations_my_dump      3    4.1 kB
migrations_my_dump_lock 1    4.1 kB
table_1                 2    4.1 kB
table_2                 2    4.1 kB
table_3                 0    4.1 kB

(20.5 kB in 5 tables)
"
`;

exports[`dump can see dumpfile contents 2`] = `
"column     type     nullable default
id         integer  false
field_1    integer  true
field_2    text     true
created_at datetime true
updated_at datetime true

index                 algorithm unique columns
table_1_field_1_index unknown   false  field_1
"
`;

exports[`dump can see dumpfile contents 3`] = `
"table                   rows bytes  columns    indexes
migrations_my_dump      3    4.1 kB 4x created none
migrations_my_dump_lock 1    4.1 kB 2x created none

Tables: 2x created, 3x similar (hidden)
Re-run with --all to show hidden rows
"
`;

exports[`dump can see dumpfile contents 4`] = `
"Columns: 5x similar (hidden)
Indexes: 1x similar (hidden)
Re-run with --all to show hidden rows
"
`;

exports[`dump dumpfile contents are valid 1`] = `
"module.exports.up = async (knex) => {
  await knex.schema.createTable(\\"table_1\\", (t) => {
    t.increments(\\"id\\");
    t.integer(\\"field_1\\");
    t.text(\\"field_2\\");
    t.datetime(\\"created_at\\");
    t.datetime(\\"updated_at\\");
    t.index([\\"field_1\\"], \\"table_1_field_1_index\\");
  });
};
module.exports.down = async (knex) => {
  await knex.schema.dropTableIfExists(\\"table_1\\");
};
"
`;

exports[`dump dumpfile contents are valid 2`] = `
"module.exports.up = async (knex) => {
  await knex.schema.createTable(\\"table_2\\", (t) => {
    t.string(\\"id\\", 255).primary();
    t.float(\\"field_1\\").notNullable();
    t.text(\\"field_2\\").defaultTo(\\"default text\\");
    t.datetime(\\"created_at\\");
    t.datetime(\\"updated_at\\");
    t.unique([\\"field_1\\"], \\"table_2_field_1_unique\\");
  });
};
module.exports.down = async (knex) => {
  await knex.schema.dropTableIfExists(\\"table_2\\");
};
"
`;

exports[`dump dumpfile contents are valid 3`] = `
"module.exports.up = async (knex) => {
  await knex.schema.createTable(\\"table_3\\", (t) => {
    t.bigInteger(\\"field_1\\");
    t.string(\\"field_2\\", 255);
    t.datetime(\\"created_at\\").notNullable().defaultTo(knex.raw(\\"CURRENT_TIMESTAMP\\"));
    t.datetime(\\"updated_at\\").notNullable().defaultTo(knex.raw(\\"CURRENT_TIMESTAMP\\"));
    t.primary([\\"field_1\\",\\"field_2\\"]);
  });
};
module.exports.down = async (knex) => {
  await knex.schema.dropTableIfExists(\\"table_3\\");
};
"
`;

exports[`dump dumpfile contents are valid 4`] = `
"{\\"id\\":1,\\"field_1\\":12,\\"field_2\\":\\"foo\\",\\"created_at\\":\\"2020-07-24 18:34:00\\"}
{\\"id\\":2,\\"field_1\\":30,\\"field_2\\":\\"bar\\",\\"created_at\\":\\"2020-07-24 18:34:00\\"}
"
`;

exports[`dump dumpfile contents are valid 5`] = `
"{\\"id\\":\\"id1\\",\\"field_1\\":12.3,\\"field_2\\":\\"foo\\",\\"created_at\\":\\"2020-07-24 18:34:00\\"}
{\\"id\\":\\"id2\\",\\"field_1\\":30.45,\\"field_2\\":\\"bar\\",\\"created_at\\":\\"2020-07-24 19:25:00\\"}
"
`;

exports[`dump loaded database has expected content 1`] = `
"column     type     nullable default
id         integer  false
field_1    integer  true
field_2    text     true
created_at datetime true
updated_at datetime true

index                 algorithm unique columns
table_1_field_1_index unknown   false  field_1
"
`;

exports[`dump loaded database has expected content 2`] = `
"column     type         nullable default
id         varchar(255) true
field_1    float        false
field_2    text         true     default text
created_at datetime     true
updated_at datetime     true

index                  algorithm unique columns
table_2_field_1_unique unknown   true   field_1
"
`;

exports[`dump loaded database has expected content 3`] = `
"column     type         nullable default
field_1    bigint       true
field_2    varchar(255) true
created_at datetime     false    CURRENT_TIMESTAMP
updated_at datetime     false    CURRENT_TIMESTAMP

There are no indexes in the table
"
`;
